### ISSUE TITLE
[Feature Request] Metadata for specs

### ISSUE THREAD

**Issue by vmoens** [pytorch/rl#2371]
> Hi there. Besides the naming, what do you think of adding some metadata for users to populate?
> 
> This could be useful for, for example, marking which keys are visual observations or temporally stacked low-dimensional observations. People commonly differentiate the type of obs through their dimensions but in this case, they have the same number of dimensions but are expected to be handled differently.
> 
> It is in general useful for subsequent algorithm implementation to parse the observations and build the network.
> 
> Like this:
> 
> ```python
> visual_obs_spec = Continuous(shape=[N, W, H])
> visual_obs_spec.meta["is_pixel"] = True
> stacked_obs_spec = Continuous(shape=[N, T, D])
> stacked_obs_spec.meta["is_pixel"] = False
> ```

@btx0424 

I like the idea but do we want more specific flags?
We could think of an API where there is a list of metadata you can put:
```python
class SpecMetadata(Enum):
    PIXELS = 0
    STATE = 1
    ACTION = 2
    ...

visual_obs_spec = Unbounded((C, H, W), dtype=torch.uint8)
state_obs_spec = Unbounded((F,), dtype=torch.float32)

visual_obs_spec.metadata.add(SpecMetadata.PIXELS)

state_obs_spec.metadata.add(SpecMetadata.STATE)

visual_obs_spec.check_metadata(SpecMetadata.STATE) # False
visual_obs_spec.check_metadata(SpecMetadata.PIXELS) # True

```

_Originally posted by @vmoens in https://github.com/pytorch/rl/issues/1818#issuecomment-2271322059_

**Comment by btx0424** [comment]
Thanks for your timely response! 

As research nowadays is shifting to more complex and realistic environments, we might need more and more complex metadata. I personally prefer more flexibility, e.g., just restricting the custom metadata to be of primitive types for serialization purposes.

**Comment by vmoens** [comment]
My idea was for this to be extensible at will, but using strings makes it clunky (it's easy to put a cap letter or a typo)

**Comment by lazizbekravshanov** [comment]
@vmoens is this still wanted? It is not on the 2026 roadmap and the thread has been quiet since 2024, so I wanted to check before building anything.

If yes, I would like to take it. Reading the thread, the open question is typo safety versus extensibility. A string enum for the well known flags might resolve both sides: members are real strings, so serialization stays trivial and the primitive types constraint @btx0424 raised holds, but access goes through something like `Meta.IS_PIXEL`, so a typo fails loudly instead of silently creating a new key. Custom keys stay plain strings for full flexibility.

The part I expect to be the real work is propagation. Metadata has to survive `clone`, `expand`, `to`, indexing and `Composite` nesting, or it silently disappears the first time a transform touches the spec. I would post a short design note here covering the API surface and that propagation table before opening a PR.

If this is stale or you would rather see effort go to something on the roadmap, happy to pick one of those up instead.

**Comment by lazizbekravshanov** [comment]
Small nudge on the above, no rush if this is not a priority.

Since posting I read through `tensor_specs.py` to see what the work actually involves, so I am ready to write the design note whenever you want it. The thing I found that shapes the design: every spec class rebuilds itself via `self.__class__(...)` with an explicit kwarg list, in `clone`, `to`, `expand`, `__getitem__` and `squeeze`/`unsqueeze`, across all the subclasses. So any new attribute silently disappears at roughly 60 sites unless it is handled. `_encode_memo_dict` already behaves that way today, which is fine for a cache and would not be fine for metadata.

That points away from threading a kwarg through every constructor and toward one choke point on the base class that copies the mapping across those reconstruction methods. Composite then works by composition, since it already clones and moves its leaves individually. Pickle round trips instance state through `__getstate__`, so a mapping of primitives needs nothing extra to survive process boundaries, which is the case that matters for `ParallelEnv` workers.

Happy to write that up properly as a design note here before any code, or to drop it if this is not something you want right now.

**Comment by vmoens** [comment]
Yes, a design note would be super useful!

Agreed, the propagation analysis is the right place to focus. Please include a concrete proposal for the “one choke point,” since reconstruction currently happens inside subclass implementations, and cover `reshape`, `unbind`, stacking heterogeneous specs, conversions between spec representations, and metadata on `Composite` itself. It would also be useful to define copy/equality semantics. Thoughts on user metadata vs standardized TorchRL metadata & extension points? 

I would not start implementation until those points are settled.

**Comment by lazizbekravshanov** [comment]
# Design note: `spec.meta` propagation

All line refs are `torchrl/data/tensor_specs.py` at `ae9719f` unless another path is given.

## 1. Proposed API

```python
spec = Unbounded(shape=(3,))
spec.meta["curriculum_stage"] = 3     # user key: any plain string
spec.meta[MetaKey.IMAGE] = True       # standard TorchRL key: string valued enum
spec.meta["image"]                    # True, enum members hash equal to their value
spec.clone().meta                     # survives reconstruction (section 2)
spec.meta["arr"] = np.zeros(3)        # TypeError, primitives only, checked at write time
```

`meta` is a small dedicated mapping class, not a bare dict, so validation lives in exactly one place, its `__setitem__`. Allowed values: `str`, `int`, `float`, `bool`, `None`. That mirrors how `Composite.names` validates at its setter (5439 to 5443) instead of the `NonTensor.example_data` approach, which accepts anything with no validation (2802).

## 2. The choke point

You are right that a naive base class wrapper intercepts nothing: `clone` (1222), `to` (1204) and `expand` (955) are abstract and overridden wholesale, base `unbind` just raises (1254), and there is no base `__getitem__` at all. There are 80 spec level `self.__class__(` reconstruction sites across the classes (82 hits total, minus 2 on `ContinuousBox` at 500 and 503). The only real template split is the reshape family: public `reshape` is defined once (1004), delegates to abstract `_reshape` (1017), and no subclass overrides the public method, with `flatten` and `unflatten` funneling there too (1020, 1032).

So the concrete proposal is `__init_subclass__` on `TensorSpec`. At class creation, for each name in a fixed list (`clone`, `to`, `expand`, `squeeze`, `unsqueeze`, `unbind`, `__getitem__`, `_reshape`, `_unflatten`), it resolves whatever the subclass actually has for that name along the MRO, skips abstract or already wrapped methods, wraps it, and sets it back on the subclass. The wrapper does one thing:

```python
out = fn(self, *args, **kwargs)
for r in (out if isinstance(out, tuple) else (out,)):
    if r is not self and isinstance(r, TensorSpec):
        self._propagate_meta(r)   # default: fill r.meta only if it is empty
return out
```

Why this survives subclass internal reconstruction: it wraps the subclass override itself, so it does not matter what the method builds inside. It stamps whatever comes back. That also catches the sites that build a different class than `self.__class__`: `MultiOneHot.to` hardcodes `MultiOneHot(` (3426) and `Composite.expand` hardcodes `Composite(` (6158). Future subclasses, including user ones, get it for free with zero edits to the 80 sites. The MRO resolution matters because `Stacked` (1496) and `StackedComposite` (6463) inherit these methods from `_LazyStackedMixin` (1261), which is not a `TensorSpec`. Precedent for `__init_subclass__` in tree: `torchrl/collectors/weight_update.py:78` and `torchrl/record/loggers/common.py:205`. The wrapper uses `functools.wraps` plus a sentinel attribute, so it stays greppable and never double wraps.

The alternative, threading a `meta=` kwarg, is roughly 80 to 90 edits, and `Composite` already shows how that ends. It hand threads `data_cls`, `step_mdp_static` and `names` today, and at this commit it has already dropped them in four places: `__getitem__` drops `data_cls` and `step_mdp_static` (5568 to 5573), `_reshape` and `empty` drop `names` (5933, 6063), and `expand` downcasts subclasses to `Composite` (6158).

Worst edge case, stated honestly: methods that return preexisting shared objects. `_LazyStackedMixin.unbind` at the stack dim returns the live `self._specs` list (1416), and `squeeze` at the stack dim returns `self._specs[0]` (1483). Stamping meta there would mutate specs the user may hold separately. So the default hook only fills when the result is a different object with empty meta, and `_LazyStackedMixin` overrides the hook to do nothing, which means stack dim `unbind` and `squeeze` return components carrying their own meta. I think that is the right semantics anyway, but it is a behavior choice and I want it visible, not hidden.

On compile: the wrapped set is construction time plumbing, and the per step methods (`rand`, `zero`, `encode`, `is_in`, `project`) are untouched. The wrapper is a plain Python call doing an identity check and a dict copy, so dynamo can inline it. I have not audited every caller of every wrapped method across the repo, so I cannot prove no wrapped call sits inside a compiled region. Flagging that as an open point rather than claiming it.

## 3. The five surfaces

| Surface | Today | Proposed | Evidence |
|---|---|---|---|
| `reshape` | one real choke point already, nobody overrides public `reshape` | wrap `_reshape` and `_unflatten`, meta is shape independent and copies through | 1004, 1017, 1020, 1032 |
| `unbind` | no funnel, every class has its own public `unbind` | covered by the wrapper; stack dim aliasing caveat above | base raise 1254, mixin 1416 |
| stacking | densify if all elements compare equal, else lazy `Stacked` | meta stays out of the equality check so it never flips dense to lazy; merge policy below | 6843, 6853, 7022, 7029 |
| conversions | build the other class directly, `mask` carries, `dtype` deliberately does not | keep conversions out of the wrapped list at first, meta does not cross representations until keys can be marked representation dependent | 2232, 3676, 4131, 4909 |
| `Composite` | shell attrs hand threaded, four sites already drop them | `meta` lives on the shell like `step_mdp_static` (5164), covered by the same wrapper, existing drop sites fixed in the same pass | 5568, 5933, 6063, 6158, plus `separates` at 5337 |

Stacking merge policy on the dense branch: today `spec0.clone()` wins everything (6849, 7028), which would silently discard the other elements' meta. Proposal: keep only the keys on which all elements agree, drop conflicting ones. That matches the intersection precedent of `StackedComposite.keys()` (7178 to 7182). On the lazy branch each element keeps its own meta inside `_specs` (1263), the way exclusive keys already work (6644).

Conversions, one open question found while auditing: `OneHot.to_categorical_spec` passes `mask` (2232) but `Categorical.to_one_hot_spec` does not (4131). I could not tell from the code whether that asymmetry is deliberate. Also the identity paths return `self` unchanged and uncloned (2243, 4137), worth documenting either way.

My lean is that conversions do not propagate meta in v1. Whether a key survives a representation change depends on the key, a layout hint does not mean the same thing on a `Categorical` as on the `OneHot` it came from, so starting conservative and adding per key propagation later seems safer than propagating everything blindly. Happy to flip that default if you disagree.

## 4. Copy and equality

Copy: every reconstruction shallow copies the dict (`dict(meta)`). Cheap given primitives only, and it avoids the `example_data` situation where payload is shared by reference through every site (2848, 2867, 2928 and friends). `to` returns `self` when nothing changes (2853, 5949), so the same dict rides along there, which is correct. Caches (`_encode_memo_dict`, `_bounds_cache`) and lock state are not meta and are not copied, matching "Locked specs will not produce locked clones" (5992 to 5995). Serialization is free: both `__getstate__` paths copy `__dict__` and filter only caches and weakrefs (672, 6279), and spec trees already cross process boundaries via stdlib pickle in `ParallelEnv` (`torchrl/envs/batched_envs.py:2139`, `EnvMetaData` at `torchrl/envs/common.py:124` and following). I would add a pickle round trip test, since `test/test_specs.py` currently has none.

Equality: recommendation is meta stays out of `__eq__`, with an opt in `spec.equals(other, include_meta=True)` for when you want it compared. Reasons:

1. `__eq__` is operational, not just for tests. `torch.stack` branches on it (6843, 7022), so meta in `__eq__` means two otherwise identical specs with different annotations silently produce a `Stacked` instead of a dense spec.
2. `__eq__` today is semantic, "do these describe the same space": `Unbounded` equals a matching `Bounded` at dtype extremes (3249 and following), `Binary` equals `Categorical` with `n == 2` (4586), bounds use `isclose` (537). Meta is not part of the space.
3. A rebuilt env will not carry hand written annotations, and the `TestEquality` pattern (`test/test_specs.py:903` and following) compares independently constructed specs. Meta in `__eq__` breaks all of that.

The codebase currently has both answers, which is worth fixing on purpose: `Composite.__eq__` includes `data_cls` but excludes `step_mdp_static` and `names` (6101 to 6109), while `NonTensor.example_data` participates (2829). The honest cost of meta out of `__eq__` is that the dense stack branch can merge specs whose meta differs, which is exactly why the agreement merge in section 3 exists.

## 5. User keys vs standardized keys, extension point

One flat mapping, two tiers by key type. Standard keys are a `MetaKey` string valued enum, user keys are plain strings in the same dict. Since `(str, Enum)` members hash and compare equal to their value, `meta[MetaKey.IMAGE]` and `meta["image"]` are the same entry, so there is no second namespace and no registration ceremony. The floor is Python 3.10 (`pyproject.toml:39`), so `enum.StrEnum` is not importable unconditionally; the in repo answer is the backport in `torchrl/data/replay_buffers/her.py:23`, with `__str__ = str.__str__` added so 3.10 and 3.11 render identically. This also fits the repo rule of typed fixed string sets over bare `str` (`CLAUDE.md:39`).

Standard key candidates, each with a consumer that exists today:

1. `step_mdp_static`. Already a bespoke bool on `Composite` (5147, 5162), produced by LLM policy specs (`torchrl/modules/llm/policies/common.py:76`), consumed via `keys(step_mdp_static_only=True)` (5823, callers `torchrl/envs/common.py:1534` to 1574), hand threaded everywhere, missing from `__eq__`, and untested. This is the flag `meta` generalizes.
2. `image`. The framework currently encodes "this is an image" as the literal key name `"pixels"`: `IMAGE_KEYS` at `torchrl/envs/transforms/_base.py:50` feeds the defaults of `ToTensorImage`, `Resize`, `GrayScale`, `CatFrames` and friends, and `VideoRecorder` defaults to it too (`torchrl/record/recorder.py:164`). A tag on the leaf spec lets transforms and the recorder stop guessing key names.
3. `role` (for locating the action leaf). `_find_action_space` locates the action by key name heuristics, with a comment admitting it breaks for anything complex (`torchrl/data/utils.py:264` to 324).
4. `action_mask_for`. `ActionMask` binds mask to action purely by default key names (`torchrl/envs/transforms/_action.py:257`).

I would not add a `register_meta_key()` registry now. It would put global mutable state in `tensor_specs.py`, which nearly everything imports, and add per key validation cost to `clone` and `expand`. It can be layered on later without breaking anything. Docs would state the reservation policy: torchrl may claim short unprefixed names for future standard keys, so users who want guaranteed permanence should prefix theirs.

## 6. Decisions I defer to you

- meta in or out of `__eq__` (I recommend out, plus opt in `equals`)
- dense stack merge policy (agreement merge vs first wins vs refusing to densify on conflict)
- whether conversions propagate meta from day one
- whether the existing `Composite` drop and downcast sites (5568, 5933, 6063, 6158, 5337) are bugs to fix in this pass or intentional
- the exact wrapped method list, in particular whether `__getitem__` belongs in it
- whether `step_mdp_static` and eventually `data_cls` migrate into `meta` as standard keys or stay as attrs
- naming, `meta` vs `metadata`

On the existing `Composite` drop sites: if you agree those are bugs rather than intentional, I would fix them in a small separate PR before this one, so the metadata change reviews clean on top of correct threading. Say the word and I will open that first.

## 7. Sequencing

Agreed on your condition: no implementation until these points are settled. Happy to iterate on any of this here first.

**Comment by lazizbekravshanov** [comment]
Small nudge on the design note above, no rush. If any of the five points needs a different shape before it is worth settling, say which and I will rework that section rather than the whole note. One thing that does not depend on the metadata design: if you confirm the four existing `Composite` drop sites are bugs, I can open that small PR first.
